fix #13222: make relativePath more robust and flexible

timotheecour opened the pull request:
* relativePath(foo) now works
* relativePath(rel, abs) and relativePath(abs, rel) now work (fixes #13222)
* relativePath, absolutePath, getCurrentDir now available in more targets (eg: ~~vm~~, nodejs etc)
* [EDIT] `isAbsolute` now works with -d:js (previously was always returning false)
see tests

timotheecour committed:
* relativePath(foo) now works
* relativePath(rel, abs) and relativePath(abs, rel) now work (fixes #13222)
* relativePath, absolutePath, getCurrentDir now available in more targets (eg: vm, nodejs etc)

timotheecour committed:
add tests

timotheecour committed:
fix bootstrap

timotheecour committed:
fixup

Araq commented:
Oh no. I don't want to expose ``getCurrentDir`` in the VM. It is a recipe for desaster builds that need a specific environment/setup. If you do ``nim c -r tests/foobar.nim`` or ``cd tests && nim c -r foobar.nim`` shouldn't be different, both should work equally well.

timotheecour committed:
remove VM getCurrentDir

timotheecour commented:
> Oh no. I don't want to expose getCurrentDir in the VM

removed it to unblock this PR; I understand the concern and this could be addressed similarly to `cpuTime` (where reproducible builds was the concern) via something analog to
```
  if optBenchmarkVM in c.config.globalOptions:
    wrap0(cpuTime, timesop)
  else:
    proc cpuTime(): float = 5.391245e-44  # Randomly chosen
    wrap0(cpuTime, timesop)
```

for users who understand that and still need the feature, but I've left this out of this PR to unblock it

timotheecour committed:
fix bug: isAbsolutePath now works with -d:js; add tests

timotheecour commented:
PTAL:
* resolve conflict
* fixed a new bug i found: isAbsolute was always returning false with `nim js` (ie on a browser) ; now it works with `nim js`; and with `nim js -d:nodejs` on posix, but not on windows, which is due to a separate, pre-existing issue, see https://github.com/nim-lang/Nim/issues/13469
* added tests

timotheecour committed:
workaround https://github.com/nim-lang/Nim/issues/13469

disruptek commented:
We talked about adding `getCurrentDir` for `bump`'s use-case.  I would use this in `nimph` and `testutils` and `golden` and `openapi`, too.  I guess I'm just a bad person.

I think you have to concede that if this isn't provided, a work-around will be developed, and that's clearly a worse option.  Can you really champion NimScript as the right tool for writing build scripts when one cannot even find the current working directory in NimScript?

Araq commented:
1. It shouldn't have been part of this PR.
2. The feature must be called ``staticGetCurrentDir`` or similar to prevent accidents like ``const x = getCurrentDir()``. And we had real **bugs** like that in the past, it's why I changed my mind on these things. I'm not evil. ;-)

Araq commented:
This patch that is really hard to review. Can't you achieve the same without rewriting the code?

timotheecour commented:
> This patch that is really hard to review. Can't you achieve the same without rewriting the code?

I agree it's hard to review but it's due to github diff view not dealing well with blocks of code being re-indented, I did not rewrite the code.

I know 2 options:

## hidden github feature: ?w=1
I've just found a really interesting solution: adding `?w=1` which I say here https://github.blog/2011-10-21-github-secrets/
As you can see in https://github.com/nim-lang/Nim/pull/13451/files?w=1 I did not rewrite code, it's just re-indentation change; this should be super helpful in other code reviews.

**caveat**: this solution is a bit blunt, see caveats here: https://github.com/timotheecour/Nim/pull/37

## git diff --color-moved-ws=allow-indentation-change --color-moved=blocks
This is the hint I added to https://nim-lang.github.io/Nim/contributing.html which helps if you locally fetch the PR, it helps by coloring differently code that was moved. 
see caveats here: https://github.com/timotheecour/Nim/pull/37

Araq commented on lib/pure/os.nim:
```diff
@@ -84,6 +84,8 @@ type
 
 include "includes/osseps"
 
+proc absolutePathInternal(path: string): string {.gcsafe.}
```
Why is this needed? And can we not forward declare it? forward decls are bad for the effect system.

timotheecour replied:
2 other options:
A1: something like https://github.com/timotheecour/Nim/pull/38 where I did the minimal changes to avoid fwd declare
it really sucks as it moves around a bunch of procs, breaks git blame/git diff history, and isn't really helping in case future PR need a dependency the other way around.

A2: https://github.com/timotheecour/Nim/pull/39 which uses `{.experimental: "codeReordering".}` 
why not? maybe it's time to use it ...

(A3: improve effect system so that it auto-infers attributes in case by solving a topological sort (analog to that in codeReordering) or fail there are cycles; maybe this would just reduce to `{.experimental: "codeReordering".}` , not sure.)

my favorite is the simple A2, or current PR, but A1 seems kind of backwards

Araq replied:
But why is it needed? Previously we managed to do without.

timotheecour replied:
absolutePathInternal is now used in `relativePath` as part of fix for #13222:
relativePath(rel, abs) and relativePath(abs, rel) now work

Araq commented on lib/pure/os.nim:
```diff
@@ -338,8 +344,8 @@ when doslikeFileSystem:
           return false
 
-proc relativePath*(path, base: string; sep = DirSep): string {.
-  noSideEffect, rtl, extern: "nos$1", raises: [].} =
```
Well previously it didn't raise an exception, if it now does, that's a breaking change.

timotheecour replied:
I've now added an explanation in changelog documenting this breaking change and in which condition it can now throw

Araq commented on changelog.md:
```diff
@@ -68,6 +68,8 @@
 - Added `minIndex`, `maxIndex` and `unzip` to the `sequtils` module.
 - Added `os.isRelativeTo` to tell whether a path is relative to another
 - Added `resetOutputFormatters` to `unittest`
+- added `os.relativePath(path)`, and made `os.relativePath` more consistent when mixing absolute
```
What's the point of this new ``os.relativePath``? To be relative against getCurrentDir()? Please remove it.

timotheecour replied:
I'll explain the rationale in some subsequent PR, for now I've removed it from this PR to unblock this

timotheecour commented:
PTAL (you can just review last 2 commits)

Araq commented on changelog.md:
```diff
@@ -27,10 +27,13 @@
-- `joinPath(a,b)` now honors trailing slashes in `b` (or `a` if `b` = "")
+- `relativePath(rel, abs)` and `relativePath(abs, rel)` used to silently give wrong results
+  (see #13222); instead they now use `getCurrentDir` to resolve those cases,
+  and this can now throw in edge cases where `getCurrentDir` throws.
```
Well getCurrentDir shouldn't raise then.

timotheecour replied:
I don't see how that's possible: raiseOSError can be triggered inside `getCurrentDir` (for reasons unrelated to this PR); and it's not theoretical but very real and happened to me a few times eg:
```
mkdir /tmp/d23 && cd /tmp/d23

nim c main.nim
./main # ok
rmdir /tmp/d23
./main
/Users/timothee/git_clone/nim/timn/tests/nim/all/main.nim(7) t10268
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/os.nim(1334) getCurrentDir
/Users/timothee/git_clone/nim/Nim_devel/lib/pure/includes/oserr.nim(94) raiseOSError
Error: unhandled exception: No such file or directory [OSError]
```

with:
```
# main.nim:
import os
echo getCurrentDir()
```

Araq commented on lib/pure/os.nim:
```diff
@@ -1351,23 +1370,26 @@ proc setCurrentDir*(newDir: string) {.inline, tags: [], noNimScript.} =
+  if isAbsolute(path): path
+  else:
+    if not root.isAbsolute:
+      raise newException(ValueError, "The specified root is not absolute: " & root)
```
The policy is to use assert/doAssert to detect invalid API usages.

timotheecour replied:
* that's **pre-existing to this PR** ( I didn't change `proc absolutePath` implementation at all) 
* this would be a breaking change (`try: someProgram() except ValueError: fallback` would stop working if someProgram calls `absolutePath` somewhere and that throws)
* even if I did change it to assert/doAssert it wouldn't make relativePath nothrow, because `absolutePathInternal` still needs `getCurrentDir` which is not nothrow as I showed in https://github.com/nim-lang/Nim/pull/13451#discussion_r385089778; so if anything this change would belong in a separate, unrelated PR
* that being said **I really don't agree** this should use assert/doAssert, we can debate somewhere else (as it's unrelated to this PR see above points) but this is an important design issue so let me elaborate:

let's suppose we used `assert`:

BUILDDIR=/tmp nim c -r -d:release main.nim
```
# main.nim
let absFile = absolutePath("bar.nim", getEnv("BUILD_DIR")).AbsolutePath
```
because of typo (user input!) BUILDDIR vs BUILD_DIR, `absFile` is now `bar.nim`, which is not an absolute path, and the whole remainder of the program is based on wrong assumptions.

Or if you instead used doAssert, then any program using absolutePath becomes non-catchable (even snippet above).
So now you're forcing the user to protect each call to AbsolutePath like this:
```nim
check(root.isAbsolute)
let root = getEnv("BUILD_DIR")
userCheck(root.isAbsolute)
let absFile = absolutePath("bar.nim", root).AbsolutePath
```
which defeats the whole point of using exceptions, which are **meant to propagate all the way up instead of having to be handled at each level**

Bad user input of this sort is an expected condition and should be handled with Exceptions, not (uncatchable) assert/doAssert. Assert/doAssert is for guaranteeing internal program logic. 
The **key distinction here is where you draw the boundary between internal logic and external input**
* private procs in a module can consider their own inputs as internal, since they're private, so they can use assert/doAssert on their inputs; everything is internal logic for those
* public procs in a public module (such as `absolutePath`) have to draw the boundary between inputs (that must be validated, using catchable exceptions) vs internal logic (which is validated using assert/doAssert); public API's can be used in any kind of contexts, unlike private / internal-only procs which operate in a controlled environment.

* for analog reasons, nim compiler code uses localError/globalError to handle user (= program being compiled) errors, but for internal logic it uses assert/doAssert/internalError; here the boundary is between input program being compiled vs internal compiler logic

## note
* D does the same: `if (!isAbsolute(baseVar)) throw new Exception("Base directory must be absolute");`  instead of assert (in D, `Exception` is catchable)
* it's the same with things like strutils.parseInt, json parsing etc: `parseInt("3.4")` throws a catchable exception on invalid input (even if it violates the contract that input should be an integer), or in this module os.getFileInfo 
* however in following case we should probably use doAssert or some other way, since no user input can be valid:
```
  proc paramStr*(i: int): TaintedString = raise newException(OSError, "paramStr is not implemented on Genode")
```

Araq commented:
You missed to update tests/js/tos.nim

timotheecour commented:
> You missed to update tests/js/tos.nim

done, PTAL (and see comments above)

Araq commented:
Gah, I screwed up the changelog cleanup...

timotheecour committed:
remove `relativePath(path)` overload for now

timotheecour committed:
add back changelog after rebase

timotheecour commented:
@Araq 
> Gah, I screwed up the changelog cleanup...

TLDR: we should rebase, not merge

I've rebased (instead of merged) to keep git history linear. Merge makes it really hard to understand individual commits as they get mixed with unrelated changes (even if this is squashed + merged in github UI, the individual commits of PR can still be seen in github UI, and are useful to follow the PR.

@disruptek
> We talked about adding getCurrentDir for bump's use-case. I would use this in nimph and testutils and golden and openapi, too. I guess I'm just a bad person.

this is now possible since https://github.com/nim-lang/Nim/pull/13813 via `--experimental:vmopsDanger`
